openedx/core/djangoapps/course_roles/helpers.py

```diff
@@ -0,0 +1,155 @@
+    return UserRole.objects.filter(
+        user=user,
+        role__permissions__name=permission_name,
+        course=course_id,
+    ).exists()
+
+
+def user_has_permission_list_course(user, permission_names, course_id):
+    """
+    Check if a user has ALL of the given permissions in a course.
+    """
+    if not use_permission_checks():
+        return False
+    return all(user_has_permission_course(user, permission_name, course_id) for permission_name in permission_names)
+
+
+def user_has_permission_list_any_course(user, permission_names, course_id):
+    """
+    Check if a user has ANY of the given permissions in a course.
+    """
+    return any(user_has_permission_course(user, permission_name, course_id) for permission_name in permission_names)
+
+
+def user_has_permission_org(user, permission_name, organization_name):
+    """
+    Check if a user has a permission for all courses in an organization.
+    """
+    if not use_permission_checks():
+        return False
+    if isinstance(user, AnonymousUser) or not isinstance(user, User):
+        return False
+    return UserRole.objects.filter(
+        user=user,
+        role__permissions__name=permission_name,
+        course__isnull=True,
+        org__name=organization_name,
+    ).exists()
+
+
+def user_has_permission_list_org(user, permission_names, organization_name):
+    """
+    Check if a user has ALL of the given permissions for all courses in an organization.
+    """
+    if not use_permission_checks():
+        return False
+    return all(
+        user_has_permission_org(user, permission_name, organization_name) for permission_name in permission_names
+    )
+
+
+def user_has_permission_course_org(user, permission_name, course_id, organization_name=None):
+    """
+    Check if a user has a permission for all courses in an organization or for a specific course.
+    """
+    if not use_permission_checks():
+        return False
+    if isinstance(user, AnonymousUser) or not isinstance(user, User):
+        return False
+    if organization_name is None:
+        course = modulestore().get_course(course_id)
+        if course:
+            organization_name = course.org
+        else:
+            return user_has_permission_course(user, permission_name, course_id)
+    return (user_has_permission_course(user, permission_name, course_id) or
+            user_has_permission_org(user, permission_name, organization_name)
+            )
+
+
+def user_has_permission_list_course_org(user, permission_names, course_id, organization_name=None):
+    """
+    Check if a user has all of the given permissions for all courses in an organization or for a specific course.
+    """
+    if not use_permission_checks():
+        return False
+    return all(
+        user_has_permission_course_org(user, permission_name, course_id, organization_name)
+        for permission_name in permission_names
+    )
+
+
+def get_all_user_permissions_for_a_course(user_id, course_id):
```

Comment:
As long as the number of possible permissions is relatively low (like it is today), one approach you could consider is to have some of your other functions call this one, and use the `@request_cached` decorator on this function only. As long as the indexes support doing that one query well, it should be fast, small, and it would only need to be run once per request, regardless of how many other pieces of code make permissions checks.

Reply:
(Assuming those permissions checks are for the same course, and that's by far the most common use case. I guess the other query pattern we'd have to worry about is, "What courses does user X have Y permission in?", which would require a different function.)